### PR TITLE
Enable Nullable Reference Types

### PR DESCRIPTION
**Pull request by dotnokato**
Enabling NRT for non-test projects and fixed warnings.
Contributes to: #271

### CODE REVIEW

File: src/MSBuild.Abstractions/MSBuild.Abstractions.csproj [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>netcoreapp3.1</TargetFramework>
+    <Nullable>enable</Nullable>
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuild.Abstractions.csproj:5]: Can you add thi once to the `Directory.Build.props` file instead of duplicating this property in every project file?

File: src/MSBuild.Abstractions/LooseVersionAssemblyLoader.cs [hunk 2]
```diff
@@ -76,7 +76,7 @@ private static Assembly TryResolveAssemblyFromPaths(AssemblyLoadContext context,
         private static Assembly LoadAndCache(AssemblyLoadContext context, string fullPath)
         {
             var assembly = context.LoadFromAssemblyPath(fullPath);
-            var name = assembly.FullName;
+            var name = AssemblyLoadContext.GetAssemblyName(fullPath).FullName;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/LooseVersionAssemblyLoader.cs:79]: why this change?

File: src/MSBuild.Abstractions/MSBuildHelpers.cs [hunk 3]
```diff
@@ -327,13 +327,13 @@ private static bool UnquoteString(ref string s)
         /// <summary>
         /// Given an optional path to MSBuild, registers an MSBuild.exe to be used for assembly resolution with this tool.
         /// </summary>
-        public static string HookAssemblyResolveForMSBuild(string msbuildPath = null)
+        public static string HookAssemblyResolveForMSBuild(string msbuildPath = "")
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildHelpers.cs:330]: please make this string nullable (`string?`) instead of using empty string

File: src/MSBuild.Abstractions/MSBuildHelpers.cs [hunk 4]
```diff
@@ -348,7 +348,7 @@ public static string HookAssemblyResolveForMSBuild(string msbuildPath = null)
         /// <summary>
         /// Given an optional path to MSBuild, finds an MSBuild path. Will query Visual Studio instances and ask for user input if there are multitple ones.
         /// </summary>
-        private static string GetMSBuildPathIfNotSpecified(string msbuildPath = null)
+        private static string GetMSBuildPathIfNotSpecified(string msbuildPath = "")
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildHelpers.cs:351]: same as above

File: src/MSBuild.Abstractions/MSBuildHelpers.cs [hunk 5]
```diff
@@ -371,7 +371,7 @@ private static string GetMSBuildPathIfNotSpecified(string msbuildPath = null)
                 // Handle selecting the version of MSBuild you want to use.
                 : SelectVisualStudioInstance(visualStudioInstances);
 
-            return instance?.MSBuildPath;
+            return instance != null ? instance.MSBuildPath : string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildHelpers.cs:374]: revert this

File: src/MSBuild.Abstractions/MSBuildWorkspaceFinder.cs [hunk 6]
```diff
@@ -22,7 +22,7 @@ public class MSBuildWorkspaceFinder
         /// <param name="searchDirectory">The base directory to search</param>
         /// <param name="workspacePath">A specific project or solution file to find</param>
         /// </summary>
-        public static (bool isSolution, string workspacePath) FindWorkspace(string searchDirectory, string workspacePath = null)
+        public static (bool isSolution, string workspacePath) FindWorkspace(string searchDirectory, string workspacePath = "")
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildWorkspaceFinder.cs:25]: please make this string nullable (`string?`) instead of using empty string

File: src/MSBuild.Abstractions/MSBuildWorkspaceFinder.cs [hunk 7]
```diff
@@ -88,7 +88,7 @@ private static string FindMatchingFile(string searchBase, Func<string, IEnumerab
         {
             if (!Directory.Exists(searchBase))
             {
-                return null;
+                return string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildWorkspaceFinder.cs:91]: revert this

File: src/MSBuild.Abstractions/MSBuildWorkspaceFinder.cs [hunk 8]
```diff
@@ -99,7 +99,7 @@ private static string FindMatchingFile(string searchBase, Func<string, IEnumerab
 
             return files.Count == 1
                 ? files[0]
-                : null;
+                : string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildWorkspaceFinder.cs:102]: revert this

File: src/MSBuild.Abstractions/MsBuildExtensions.cs [hunk 9]
```diff
@@ -8,11 +8,11 @@ namespace MSBuild.Abstractions
 {
     public static class MsBuildExtensions
     {
-        public static XmlElement GetXml(this ProjectItemElement projectItem)
+        public static XmlElement? GetXml(this ProjectItemElement projectItem)
         {
-            return (XmlElement)GetPropertyValue(projectItem, "XmlElement");
+            return GetPropertyValue(projectItem, "XmlElement") as XmlElement;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MsBuildExtensions.cs:13]: When will this not be an `XmlElement`? Why not just explicitly cast it?

File: src/MSBuild.Abstractions/MsBuildExtensions.cs [hunk 10]
```diff
@@ -27,17 +27,17 @@ static object GetPropertyValue(object obj, string propertyName)
                       string.Format("Couldn't find property {0} in type {1}", propertyName, objType.FullName));
                 }
 
-                return propInfo.GetValue(obj, null);
+                return propInfo?.GetValue(obj, null);
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MsBuildExtensions.cs:30]: `?.` is unnecessary, we already check for null up above.

File: src/MSBuild.Abstractions/MsBuildExtensions.cs [hunk 11]
```diff
@@ -27,17 +27,17 @@ static object GetPropertyValue(object obj, string propertyName)
...
             }
 
-            static PropertyInfo GetPropertyInfo(Type type, string propertyName)
+            static PropertyInfo? GetPropertyInfo(Type? type, string propertyName)
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MsBuildExtensions.cs:33]: can we make `type` non-null and then check that its not null on line 23 before calling `GetPropertyInfo`?

File: src/MSBuild.Abstractions/NugetHelpers.cs [hunk 12]
```diff
@@ -40,7 +40,7 @@ static string GetVersionFromQueryResponse(Stream result)
                     }
                 }
 
-                return null;
+                return string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/NugetHelpers.cs:43]: revert this and make the return type `Task<string?>`

File: src/MSBuild.Conversion.Package/Packages.cs [hunk 13]
```diff
@@ -5,22 +5,22 @@ public class PackagesConfigPackage
         /// <summary>
         /// Name of the package.
         /// </summary>
-        public string ID { get; set; }
+        public string ID { get; set; } = string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Conversion.Package/Packages.cs:8]: I would just make these nullable as this is a POCO.

File: src/MSBuild.Conversion.Project/ProjectRootElementExtensions.cs [hunk 14]
```diff
@@ -156,7 +156,7 @@ public static IProjectRootElement RemoveOrUpdateItems(this IProjectRootElement p
                     else if (ProjectItemHelpers.IsReferenceConvertibleToPackageReference(item))
                     {
                         string packageName = NugetHelpers.FindPackageNameFromReferenceName(item.Include);
-                        string version = null;
+                        string version = string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Conversion.Project/ProjectRootElementExtensions.cs:159]: just use null instead of `string.Empty`

File: src/MSBuild.Conversion.SDK/TargetFrameworkHelper.cs [hunk 15]
```diff
@@ -19,7 +19,7 @@ public static class TargetFrameworkHelper
         public static string FindHighestInstalledTargetFramework(bool usePreviewSDK)
         {
             // Finds SDK path
-            string sdkPath = null;
+            string sdkPath = string.Empty;
```
Comment (jmarolf) [inline src/MSBuild.Conversion.SDK/TargetFrameworkHelper.cs:22]: use `null` instead of `string.Empty`

**Comment by jmarolf** [comment]
Looks good, in general I don't see a reason to prefer `string.Empty` to `null`. Both represent invalid states, but only one (`null`) is tracked by the compiler. I would prefer it if we just used `null` to represent a failure case.

File: src/MSBuild.Abstractions/LooseVersionAssemblyLoader.cs (hunk 2 above)
Comment (dotnokato) [reply src/MSBuild.Abstractions/LooseVersionAssemblyLoader.cs:79]: My logic was:
`context.LoadFromAssemblyPath(fullPath);` returns `Assembly` instance in which `string? FullName` is nullable
whereas `AssemblyLoadContext.GetAssemblyName(fullPath)` returns `AssemblyName` instance where `string FullName` is marked as non-nullable.

Shall I revert the change?

Comment (jmarolf) [reply src/MSBuild.Abstractions/LooseVersionAssemblyLoader.cs:79]: I think its ok to keep it. I'll ask the CoreFx folks what horrible back-compat story they must have for why `Assembly.FullName` can return null but `AssemblyLoadContext.GetAssemblyName("").FullName` cannot. I assume that `AssemblyLoadContext.GetAssemblyName` throws if it cannot file the assembly name where as `Assembly.FullName` returns null and they need to keep this around for backcompat.

File: src/MSBuild.Abstractions/MsBuildExtensions.cs (hunk 11 above)
Comment (dotnokato) [reply src/MSBuild.Abstractions/MsBuildExtensions.cs:33]: I marked `type` as nullable because we call `type.BaseType;` on line 40 which returns nullable `Type? BaseType` and I was getting warning.
Let me rework that.

File: src/MSBuild.Abstractions/MsBuildExtensions.cs (hunk 9 above)
Comment (dotnokato) [reply src/MSBuild.Abstractions/MsBuildExtensions.cs:13]: This was a result of marking `GetPropertyValue `as nullable.
I'll revert that.

**Comment by dotnokato** [comment]
@jmarolf Thanks for the review and comments.
I've reworked the changes to prefer `null `over `string.Empty` as suggested.
Please let me know if there is anything else that catches your eye.

File: src/MSBuild.Abstractions/MsBuildExtensions.cs [hunk 16]
```diff
@@ -16,30 +16,37 @@ static object GetPropertyValue(object obj, string propertyName)
...
+                      string.Format("Null value for property {0} in type {1}", propertyName, objType.FullName));
+                }
+                return propertyValue;
             }
 
-            static PropertyInfo GetPropertyInfo(Type type, string propertyName)
+            static PropertyInfo? GetPropertyInfo(Type type, string propertyName)
             {
-                PropertyInfo propInfo;
+                PropertyInfo? propInfo;
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MsBuildExtensions.cs:41]: >propInfo [](start = 30, length = 8)

I don't think the intention is that this ever returns null. perhaps throw and exception in that case? I would rather have an exception for this than needing to track it down through other means

File: src/MSBuild.Abstractions/MSBuildHelpers.cs [hunk 17]
```diff
@@ -217,9 +217,10 @@ public static bool HasWPFOrWinForms(ProjectPropertyGroupElement propGroup)
         /// <summary>
         /// Checks if a given TFM is not .NET Framework.
         /// </summary>
-        public static bool IsNotNetFramework(string tfm) =>
-            !tfm.ContainsIgnoreCase(MSBuildFacts.NetcoreappPrelude)
-            && !tfm.ContainsIgnoreCase(MSBuildFacts.NetstandardPrelude);
+        public static bool IsNotNetFramework(string? tfm) =>
+            tfm is null
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/MSBuildHelpers.cs:221]: I would not allow this to accept `null`. I don't think this check should succeed becasue we passed in null.

File: src/MSBuild.Abstractions/ProjectItemHelpers.cs [hunk 18]
```diff
@@ -112,7 +112,7 @@ public static class ProjectItemHelpers
         /// <summary>
         /// Checks if an item is an explicit System.ValueTuple and if the given TFM correspondes with an in-box System.ValueTuple type.
         /// </summary>
-        public static bool IsExplicitValueTupleReferenceThatCanBeRemoved(ProjectItemElement item, string tfm) =>
+        public static bool IsExplicitValueTupleReferenceThatCanBeRemoved(ProjectItemElement item, string? tfm) =>
```
Comment (jmarolf) [inline src/MSBuild.Abstractions/ProjectItemHelpers.cs:115]: I would not allow this to be `null`. I will admit that in most cases comparing anything to `null` _should_ result in false. But the code here expects to be given a valid TFM.

File: src/try-convert/Program.cs [hunk 19]
```diff
@@ -35,7 +35,7 @@ private static async Task<int> Main(string[] args)
             return await parser.InvokeAsync(args.Length > 0 ? args : new string[] { "-h" }).ConfigureAwait(false);
         }
 
-        public static int Run(string project, string workspace, string msbuildPath, string tfm, bool allowPreviews, bool diffOnly, bool noBackup)
+        public static int Run(string project, string workspace, string? msbuildPath, string tfm, bool allowPreviews, bool diffOnly, bool noBackup)
```
Comment (jmarolf) [inline src/try-convert/Program.cs:38]: so, technically all the strings in this method can be given null and we need to be resilient to it so I would update all of them to be `string?`. I would also update the `AddOption` clauses above to return `new Argument<string?>()` instead of `new Argument<string>()`

**Comment by jmarolf** [comment]
@dotnokato look great! I resolved all the comments you fixed and left a few nits. fixup those last 4 things and I think this is good to merge

File: src/MSBuild.Abstractions/MsBuildExtensions.cs (hunk 16 above)
Comment (dotnokato) [reply src/MSBuild.Abstractions/MsBuildExtensions.cs:36]: I have moved the null check to within `GetPropertyInfo` and marked it as not nullable.

File: src/MSBuild.Abstractions/MsBuildExtensions.cs [hunk 20]
```diff
@@ -16,30 +16,38 @@ static object GetPropertyValue(object obj, string propertyName)
...
                 var propInfo = GetPropertyInfo(objType, propertyName);
-                if (propInfo == null)
+
+                var propertyValue = propInfo.GetValue(obj, null);
+                if (propertyValue == null)
                 {
-                    throw new ArgumentOutOfRangeException("propertyName",
-                      string.Format("Couldn't find property {0} in type {1}", propertyName, objType.FullName));
+                    throw new ArgumentOutOfRangeException(nameof(propertyName),
+                      string.Format("Null value for property {0} in type {1}", propertyName, objType.FullName));
```
[the remark this answers is not in the document]
Comment (dotnokato) [reply src/MSBuild.Abstractions/MsBuildExtensions.cs:29]: If I restore the original message here we will have two places throwing exception with the same message. Do we want that?

**Comment by dotnokato** [comment]
@jmarolf I've pushed next set of changes.
I guess this leaves us with only https://github.com/dotnet/try-convert/pull/272#discussion_r404368085 unresolved. Let me know if you want me to revert to the original exception message.

**Comment by jmarolf** [comment]
yeah, I don't think the exception is worth holding things up over. I queued the job that failed to re-run. If that passed I am going to merge this PR. Thanks so much for all your work!
